packages/firestore/src/local/query_engine.ts

```diff
@@ -109,20 +133,111 @@ export class QueryEngine {
+
     return this.performQueryUsingIndex(transaction, query)
-      .next(result =>
-        result
-          ? result
-          : this.performQueryUsingRemoteKeys(
-              transaction,
-              query,
-              remoteKeys,
-              lastLimboFreeSnapshotVersion
-            )
-      )
-      .next(result =>
-        result ? result : this.executeFullCollectionScan(transaction, query)
+      .next(result => {
+        queryResult.result = result;
+      })
+      .next(() => {
+        if (queryResult.result) {
+          return;
+        }
+        return this.performQueryUsingRemoteKeys(
+          transaction,
+          query,
+          remoteKeys,
+          lastLimboFreeSnapshotVersion
+        ).next(result => {
+          queryResult.result = result;
+        });
+      })
+      .next(() => {
+        if (queryResult.result) {
+          return;
+        }
+        const context = new QueryContext();
+        return this.executeFullCollectionScan(transaction, query, context).next(
+          result => {
+            queryResult.result = result;
+            if (this.indexAutoCreationEnabled) {
+              return this.createCacheIndexes(
+                transaction,
+                query,
+                context,
+                result.size
+              );
+            }
+          }
+        );
+      })
+      .next(() => queryResult.result!);
+  }
+
+  createCacheIndexes(
+    transaction: PersistenceTransaction,
+    query: Query,
+    context: QueryContext,
+    resultSize: number
+  ): PersistencePromise<void> {
+    if (context.documentReadCount < this.indexAutoCreationMinCollectionSize) {
+      if (getLogLevel() <= LogLevel.DEBUG) {
+        logDebug(
+          'QueryEngine',
+          'SDK will not create cache indexes for query:',
+          stringifyQuery(query),
+          'since it only creates cache indexes for collection contains',
+          'more than or equal to',
+          this.indexAutoCreationMinCollectionSize,
+          'documents'
+        );
+      }
+      return PersistencePromise.resolve();
+    }
+
+    if (getLogLevel() <= LogLevel.DEBUG) {
+      logDebug(
+        'QueryEngine',
+        'Query:',
+        stringifyQuery(query),
+        'scans',
+        context.documentReadCount,
+        'local documents and returns',
+        resultSize,
+        'documents as results.'
+      );
+    }
+
+    if (
+      context.documentReadCount >
+      this.relativeIndexReadCostPerDocument * resultSize
+    ) {
+      if (getLogLevel() <= LogLevel.DEBUG) {
+        logDebug(
+          'QueryEngine',
+          'The SDK decides to create cache indexes for query:',
+          stringifyQuery(query),
+          'as using cache indexes may help improve performance.'
+        );
+      }
+      return this.indexManager.createTargetIndexes(
+        transaction,
+        queryToTarget(query)
       );
+    } else {
```

Comment:
That _is_ true that this log message is somewhat superfluous; however, the other SDKs include this specific log message. Do you think we should remove it from Android and iOS too? Personally, I _like_ having this log message even though it can be deduced because the absence of a log message is less clear than a direct log message IMO. But I'm fine to remove it if you think the code is better without it.

Reply:
I think Cheryl is removing this for Android/iOS, let's remove it here too.

Reply:
Done.